api/server/docker.go

```diff
@@ -169,79 +226,123 @@ func (d *driver) mountpath(name string) string {
 }
 
+func (d *driver) getConn() (*grpc.ClientConn, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.conn == nil {
+		var err error
+		d.conn, err = grpcserver.Connect(
+			d.sdkUds,
+			[]grpc.DialOption{grpc.WithInsecure()})
+		if err != nil {
+			return nil, fmt.Errorf("Failed to connect to gRPC handler: %v", err)
+		}
+	}
+	return d.conn, nil
+}
+
 func (d *driver) create(w http.ResponseWriter, r *http.Request) {
 	method := "create"
+	ctx := r.Context()
+
+	// decode request
 	request, err := d.decode(method, w, r)
 	if err != nil {
 		return
 	}
 
+	// attach token in context metadata
+	ctx = d.attachToken(ctx, request)
+
+	// get spec for volume creation
 	specParsed, spec, locator, source, name := d.SpecFromString(request.Name)
 	d.logRequest(method, name).Infoln("")
-	// If we fail to find the volume, create it.
-	if _, err = d.volFromName(name); err != nil {
-		v, err := volumedrivers.Get(d.name)
+	if !specParsed {
+		spec, locator, source, err = d.SpecFromOpts(request.Opts)
 		if err != nil {
 			d.errorResponse(method, w, err)
 			return
 		}
-		if !specParsed {
-			spec, locator, source, err = d.SpecFromOpts(request.Opts)
-			if err != nil {
-				d.errorResponse(method, w, err)
-				return
-			}
-		}
-		if locator == nil {
-			locator = &api.VolumeLocator{}
-		}
-		locator.Name = name
-		if source != nil && len(source.Parent) != 0 {
-			vol, err := d.volFromName(source.Parent)
-			if err != nil {
-				d.errorResponse(method, w, err)
-				return
-			}
-			if _, err = v.Snapshot(vol.Id,
-				false,
-				&api.VolumeLocator{Name: name},
-				false,
-			); err != nil {
-				d.errorResponse(method, w, err)
-				return
-			}
-		} else if _, err := v.Create(
-			locator,
-			nil,
-			spec,
-		); err != nil && err != volume.ErrExist {
-			d.errorResponse(method, w, err)
-			return
-		}
 	}
+
+	// get grpc connection
+	conn, err := d.getConn()
+	if err != nil {
+		d.errorResponse(method, w, err)
+		return
+	}
+
+	// clone if exists, create otherwise
+	spec.VolumeLabels = locator.VolumeLabels
+	volumes := api.NewOpenStorageVolumeClient(conn)
+	if source != nil && len(source.Parent) != 0 {
+		// clone
+		_, err = volumes.Clone(ctx, &api.SdkVolumeCloneRequest{
+			Name:     name,
+			ParentId: source.Parent,
+		})
+	} else {
+		// create
+		_, err = volumes.Create(ctx, &api.SdkVolumeCreateRequest{
+			Name: name,
+			Spec: spec,
+		})
+	}
+	if err != nil {
+		d.errorResponse(method, w, err)
+		return
+	}
+
 	json.NewEncoder(w).Encode(&volumeResponse{})
 }
 
 func (d *driver) remove(w http.ResponseWriter, r *http.Request) {
 	method := "remove"
+	ctx := r.Context()
+
 	request, err := d.decode(method, w, r)
 	if err != nil {
 		return
 	}
 
-	v, err := volumedrivers.Get(d.name)
+	// attach token in context metadata
+	ctx = d.attachToken(ctx, request)
+
+	// get spec for deletion
+	specParsed, _, _, _, name := d.SpecFromString(request.Name)
+	d.logRequest(method, name).Infoln("")
+	if !specParsed {
+		_, _, _, err = d.SpecFromOpts(request.Opts)
```

Comment:
nit: if you are not using the opts anywhere for remove, there is no reason to error out here?